### ISSUE TITLE
FBE not working correctly on pre-U devices

### ISSUE THREAD

**Issue by BlackMesa123** [BlackMesa123/UN1CA#2]
[File-Based Encryption](https://source.android.com/docs/security/features/encryption/file-based) has certainly become a standard in Android nowadays, that's why there shouldn't really be a reason to get rid of it. On Android 14, either Google nor Samsung have certainly changed stuff under the hood, as it seems like pre-U devices aren't able to decrypt userdata correctly once a PIN/password/pattern has been set by the user. Once the device gets rebooted after setting up a PIN/password/pattern, the OS will ask for it to decrypt the userdata in order to continue the boot process. However when doing so, the SystemUI app will crash, preventing this from happening. Giving a look at the logs we can see this is happening outside of SystemUI app tho:
```
11-29 18:08:19.559  4967  5029 E AndroidRuntime: FATAL EXCEPTION: AsyncTask #1
11-29 18:08:19.559  4967  5029 E AndroidRuntime: Process: com.android.systemui, PID: 4967
11-29 18:08:19.559  4967  5029 E AndroidRuntime: java.lang.RuntimeException: An error occurred while executing doInBackground()
11-29 18:08:19.559  4967  5029 E AndroidRuntime: 	at android.os.AsyncTask$4.done(AsyncTask.java:415)
11-29 18:08:19.559  4967  5029 E AndroidRuntime: 	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:381)
11-29 18:08:19.559  4967  5029 E AndroidRuntime: 	at java.util.concurrent.FutureTask.setException(FutureTask.java:250)
11-29 18:08:19.559  4967  5029 E AndroidRuntime: 	at java.util.concurrent.FutureTask.run(FutureTask.java:269)
11-29 18:08:19.559  4967  5029 E AndroidRuntime: 	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:305)
11-29 18:08:19.559  4967  5029 E AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
11-29 18:08:19.559  4967  5029 E AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
11-29 18:08:19.559  4967  5029 E AndroidRuntime: 	at java.lang.Thread.run(Thread.java:1012)
11-29 18:08:19.559  4967  5029 E AndroidRuntime: Caused by: android.os.ServiceSpecificException:  (code -22)
11-29 18:08:19.559  4967  5029 E AndroidRuntime: 	at android.os.Parcel.createExceptionOrNull(Parcel.java:3083)
11-29 18:08:19.559  4967  5029 E AndroidRuntime: 	at android.os.Parcel.createException(Parcel.java:3053)
11-29 18:08:19.559  4967  5029 E AndroidRuntime: 	at android.os.Parcel.readException(Parcel.java:3036)
11-29 18:08:19.559  4967  5029 E AndroidRuntime: 	at android.os.Parcel.readException(Parcel.java:2978)
11-29 18:08:19.559  4967  5029 E AndroidRuntime: 	at com.android.internal.widget.ILockSettings$Stub$Proxy.checkCredential(ILockSettings.java:2087)
11-29 18:08:19.559  4967  5029 E AndroidRuntime: 	at com.android.internal.widget.LockPatternUtils.checkCredential(LockPatternUtils.java:617)
11-29 18:08:19.559  4967  5029 E AndroidRuntime: 	at com.android.internal.widget.LockPatternChecker$2.doInBackground(LockPatternChecker.java:110)
11-29 18:08:19.559  4967  5029 E AndroidRuntime: 	at com.android.internal.widget.LockPatternChecker$2.doInBackground(LockPatternChecker.java:104)
11-29 18:08:19.559  4967  5029 E AndroidRuntime: 	at android.os.AsyncTask$3.call(AsyncTask.java:394)
11-29 18:08:19.559  4967  5029 E AndroidRuntime: 	at java.util.concurrent.FutureTask.run(FutureTask.java:264)
11-29 18:08:19.559  4967  5029 E AndroidRuntime: 	... 4 more
11-29 18:08:19.559  4967  5029 E AndroidRuntime: Caused by: android.os.RemoteException: Remote stack trace:
11-29 18:08:19.559  4967  5029 E AndroidRuntime: 	at android.os.Parcel.createExceptionOrNull(Parcel.java:3083)
11-29 18:08:19.559  4967  5029 E AndroidRuntime: 	at android.os.Parcel.createException(Parcel.java:3053)
11-29 18:08:19.559  4967  5029 E AndroidRuntime: 	at android.os.Parcel.readException(Parcel.java:3036)
11-29 18:08:19.559  4967  5029 E AndroidRuntime: 	at android.os.Parcel.readException(Parcel.java:2978)
11-29 18:08:19.559  4967  5029 E AndroidRuntime: 	at android.os.IVold$Stub$Proxy.unlockUserKey(IVold.java:2696)
11-29 18:08:19.559  4967  5029 E AndroidRuntime: 
```

Code -22 in keymaster_defs.h corresponds to [KM_ERROR_KEY_EXPORT_OPTIONS_INVALID](https://android.googlesource.com/platform/hardware/libhardware/+/d783cabd4d9bddb4b83f2dd38300b7598bb58b24/include_all/hardware/keymaster_defs.h#463), so most likely the issue is either in vold or the keymaster TA.

Full debug logs, both pre-crypto (setting up PIN) and post-crypto (fresh boot):
[logs.zip](https://github.com/BlackMesa123/UN1CA/files/13534021/logs.zip)

**Comment by BlackMesa123** [comment]
This has been fixed in A52s with 6c053fe0b4166777a3b30e42cfbd9114ece98725, waiting for M52 testers to confirm the same fix works there as well.

**Comment by BlackMesa123** [comment]
Confirmed to be working on M52 aswell, closing this issue as solved.
